File: kth-largest-sum-in-a-binary-tree/kth-largest-sum-in-a-binary-tree.java

class Solution {
    public long kthLargestLevelSum(TreeNode root, int k) {
        PriorityQueue<Long> pq = new PriorityQueue<>(Collections.reverseOrder());
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            long sz = q.size(), sum = 0;
            for(int i=0; i<sz; i++){
                TreeNode node = q.remove();
                sum += node.val;
                if(node.left!=null) q.add(node.left);
                if(node.right!=null) q.add(node.right);
            }
            pq.add(sum);
        }
        if(pq.size() < k) return -1;
        for(int i=0; i<k-1; i++) pq.remove();
        return pq.peek();
    }
}
// Kth Largest sum in a binary tree

File: kth-largest-sum-in-a-binary-tree/kth-largest-sum-in-a-binary-tree.cpp

class Solution {
public:
    long long kthLargestLevelSum(TreeNode* root, int k) {
        priority_queue<long, vector<long>, greater<long> > pq;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            long n = q.size(), sum = 0;
            for(int i=0; i<n; i++){
                TreeNode *node = q.front(); q.pop();
                sum += node->val;
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
            pq.push(sum);
            if(pq.size() > k) pq.pop();
        }
        if(pq.size() < k) return -1;
        return pq.top();
    }
};